tests/diff-folded.rs

```diff
@@ -0,0 +1,159 @@
+                let mut f = File::create(expected_result_file).unwrap();
+                diff_folded::from_files(&options, &infile1, &infile2, &mut f)?;
+                fs::metadata(expected_result_file).unwrap()
+            } else {
+                return Err(e.into());
+            }
+        }
+    };
+
+    let expected_len = metadata.len() as usize;
+    let mut result = Cursor::new(Vec::with_capacity(expected_len));
+    let return_value = diff_folded::from_files(&options, infile1, infile2, &mut result)?;
+    let expected = BufReader::new(File::open(expected_result_file).unwrap());
+    // write out the expected result to /tmp for easy restoration
+    result.set_position(0);
+    let rand: u64 = rand::random();
+    let tm = std::env::temp_dir().join(format!("test-{}.svg", rand));
+    if fs::write(&tm, result.get_ref()).is_ok() {
+        eprintln!("test output in {}", tm.display());
+    }
+    // and then compare
+    result.set_position(0);
+    compare_results(result, expected, expected_result_file);
+    Ok(return_value)
+}
+
+fn compare_results<R, E>(result: R, expected: E, expected_file: &str)
+where
+    R: BufRead,
+    E: BufRead,
+{
+    let result_lines: Result<Vec<String>, _> = result.lines().collect();
+    let mut result_lines = result_lines.unwrap();
+    let expected_lines: Result<Vec<String>, _> = expected.lines().collect();
+    let mut expected_lines = expected_lines.unwrap();
+
+    assert_eq!(
+        result_lines.len(),
+        expected_lines.len(),
+        "\nresult has {} lines, expected {} lines",
+        result_lines.len(),
+        expected_lines.len()
+    );
+
+    result_lines.sort_unstable();
+    expected_lines.sort_unstable();
+
+    for (line_num, (result_line, expected_line)) in result_lines
+        .into_iter()
+        .zip(expected_lines.into_iter())
+        .enumerate()
+    {
+        assert_eq!(
+            result_line, expected_line,
+            "\n{}:{}",
+            expected_file, line_num
+        );
+    }
+}
+
+fn test_diff_folded_logs<F>(infile1: &str, infile2: &str, asserter: F)
+where
+    F: Fn(&Vec<testing_logger::CapturedLog>),
+{
+    test_diff_folded_logs_with_options(infile1, infile2, asserter, Default::default());
+}
+
+fn test_diff_folded_logs_with_options<F>(
+    infile1: &str,
+    infile2: &str,
+    asserter: F,
+    mut options: Options,
+) where
+    F: Fn(&Vec<testing_logger::CapturedLog>),
+{
+    testing_logger::setup();
+    let r1 = BufReader::new(File::open(infile1).unwrap());
+    let r2 = BufReader::new(File::open(infile2).unwrap());
+    let sink = io::sink();
+    let _ = diff_folded::from_readers(&mut options, r1, r2, sink);
+    testing_logger::validate(asserter);
+}
+
+#[test]
+fn diff_folded_default() {
+    let infile1 = "./tests/data/diff-folded/before.txt";
+    let infile2 = "./tests/data/diff-folded/after.txt";
+    let expected_result_file = "./tests/data/diff-folded/results/default.txt";
```

Comment:
And just to be paranoid, you've checked all of the `results/` files against the version produced by upstream, right? I almost wonder whether we should have a script that re-generates all of the things in `results` from upstream.. But that's for another time.

Reply:
The files in `/results` were actually produced by upstream, so yes.